play-validations/memory-footprint/src/main/java/com/google/wear/watchface/dfx/memory/AmbientMemoryFootprintCalculator.java

```diff
@@ -78,17 +74,37 @@ class AmbientMemoryFootprintCalculator {
         this.ambientConfigValue = VariantConfigValue.ambient(evaluationSettings);
         this.document = document;
         this.resourceMemoryMap = resourceMemoryMap;
-        this.resourceCollector =
-                new WatchFaceResourceCollector(document, resourceMemoryMap, evaluationSettings);
         this.evaluationSettings = evaluationSettings;
+        this.resourceCollector = createResourceCollector();
+    }
 
-        if (evaluationSettings.deduplicateAmbient()) {
-            computeResourceDedupMap();
+    private WatchFaceResourceCollector createResourceCollector() {
+        if (!evaluationSettings.deduplicateAmbient()) {
+            return new WatchFaceResourceCollector(document, resourceMemoryMap, evaluationSettings);
         }
+        Map<String, String> dedupMap =
+                computeResourceDedupMap(resourceMemoryMap, evaluationSettings);
+        return new WatchFaceResourceCollector(document, resourceMemoryMap, evaluationSettings) {
+
+            @Override
+            Set<String> collectResources(Node currentNode, VariantConfigValue variant) {
+                return super.collectResources(currentNode, variant).stream()
+                        .map(dedupMap::get)
+                        .collect(toSet());
+            }
+
+            @Override
+            Set<String> collectResources(Node node) {
+                return super.collectResources(node).stream().map(dedupMap::get).collect(toSet());
+            }
+        };
     }
```

Comment:
This hides the logic of wether or not we should deduplicate resources in the `WatchFaceResourceCollector`, which is then passed to the `DynamicNodePerConfigurationFootprintCalculator` class. Previously, that information was scoped only to the `AmbientMemoryFootprintCalculator` class, but with this change, we need it further along the pipeline and encapsulating it in the `WatchFaceResourceCollector` seems to be the cleanest way to pass it down.